File: src/main/java/io/binac/leetcode/TrappingRainWater.java
package io.binac.leetcode;

/**
 * Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.
 * <p>
 * <p>http://www.leetcode.com/static/images/problemset/rainwatertrap.png
 * <p>The above elevation map is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped. Thanks Marcos for contributing this image!
 * <p>
 * <p>Example:
 * <blockquote><pre>
 *     Input: [0,1,0,2,1,0,1,3,2,1,2,1]
 *     Output: 6
 * </blockquote></pre>
 */
public class TrappingRainWater {
    public static class Solution1 {
        public int trap(int[] height) {
            if (height.length < 3) return 0;

            int leftMax[] = new int[height.length];
            int rightMax[] = new int[height.length];
            for (int i = 1; i < height.length; ++i)
                leftMax[i] = Math.max(leftMax[i - 1], height[i - 1]);
            for (int i = height.length - 2; i >= 0; --i)
                rightMax[i] = Math.max(rightMax[i + 1], height[i + 1]);

            int result = 0;
            for (int i = 1; i < height.length - 1; ++i) {
                final int h = Math.min(leftMax[i], rightMax[i]) - height[i];
                if (h > 0) result += h;
            }
            return result;
        }
    }

    public static class Solution2 {
        public int trap(int[] height) {
            if (height.length < 3) return 0;

            int result = 0;
            int stack[] = new int[height.length];
            int top = -1;
            for (int i = 0; i < height.length; ++i) {
                while (top != -1 && height[i] > height[stack[top]]) {
                    int j = stack[top--];
                    if (top == -1) break;
                    int leftBound = stack[top];
                    int length = i - leftBound - 1;
                    int h = Math.min(height[i], height[leftBound]) - height[j];
                    result += length * h;
                }
                stack[++top] = i;
            }
            return result;
        }
    }

    public static class Solution3 {
        public int trap(int[] height) {
            if (height.length < 3) return 0;

            int result = 0;
            int left = 0, right = height.length - 1;
            int leftMax = 0, rightMax = 0;
            while (left < right) {
                if (height[left] < height[right]) {
                    if (height[left] < leftMax)
                        result += leftMax - height[left];
                    else
                        leftMax = height[left];
                    ++left;
                } else {
                    if (height[right] < rightMax)
                        result += rightMax - height[right];
                    else
                        rightMax = height[right];
                    --right;
                }
            }
            return result;
        }
    }
}

File: python3/leetcodepy/trapping_rain_water.py
"""
Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.



Example 1:

https://assets.leetcode.com/uploads/2018/10/22/rainwatertrap.png
Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
Output: 6
Explanation: The above elevation map (black section) is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped.

Example 2:

Input: height = [4,2,0,3,2,5]
Output: 9


Constraints:

n == height.length
0 <= n <= 3 * 10^4
0 <= height[i] <= 10^5
"""
from typing import List


class Solution1:
    def trap(self, height: List[int]) -> int:
        size = len(height)
        if size < 3:
            return 0
        result = 0
        left_max = [0] * size
        for i in range(1, size):
            left_max[i] = max(left_max[i - 1], height[i - 1])
        right_max = [0] * size
        for i in range(size - 2, 0, -1):
            right_max[i] = max(right_max[i + 1], height[i + 1])
            h = min(left_max[i], right_max[i]) - height[i]
            if h > 0:
                result += h
        return result


class Solution2:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        result = 0
        size = len(height)
        stack = [0]
        for i in range(1, size):
            while stack and height[i] > height[stack[-1]]:
                j = stack.pop()
                if not stack:
                    break
                left_bound = stack[-1]
                length = i - left_bound - 1
                h = min(height[left_bound], height[i]) - height[j]
                result += h * length
            stack.append(i)
        return result


class Solution3:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        result = 0
        i, j = 0, len(height) - 1
        left_max, right_max = 0, 0
        while i < j:
            if height[i] <= height[j]:
                if left_max <= height[i]:
                    left_max = height[i]
                else:
                    result += left_max - height[i]
                i += 1
            else:
                if right_max <= height[j]:
                    right_max = height[j]
                else:
                    result += right_max - height[j]
                j -= 1
        return result
